custom_components/localtuya/__init__.py

```diff
@@ -34,6 +34,17 @@
 
 UNSUB_LISTENER = "unsub_listener"
 
+DEVICE_SCHEMA = {
```

Comment:
Did not get what you're saying here, maybe it's because it's late...

Reply:
Just like we provide `PLATFORM_SCHEMA` for each platform to define what you are allowed to write in `configuration.yaml`, we now need to remove all of them and add a schema to then component that verifies the same thing, but on "domain" level. The component schema basically validated:

```yaml
localtuya:
  what_you_can_write_here
```

We basically want to ensure that `what_you_can_write_here` is a list of `DEVICE_SCHEMA` items. A thing that kinda passed me by is that we need to do some kind of special validation for the entities, so you can't write configuration for a `light` entity under `switch` for instance. Not sure how to do that, will have to investigate a bit. It must be done dynamically though.

Reply:
Yeah, this is where I really need your help because the Schema mechanism is not clear to me. I still don't get where it is used and how the defaults are being populated, in particular when importing from YAML...

Reply:
Yeah, it's a bit tricky since it's part of Thea architecture of how a component works. But the schema defined by `CONFIG_SCHEMA` is basically what is used to validate:

```yaml
localtuya:
  ...
```

And is always defined in `_init__.py`. For platforms, `PLATFORM_SCHEMA` is used to validate:

```yaml
sensor:
  - platform: localtuya
```

And must in this case be defined in `localtuya/sensor.py`.

The schema used by a config entry is on the other hand completely determined by the config flow. So there's no schema definition like for YAML.

I'm working on a schema for this, it's a bit tricky but I have an idea. Hopefully we can share some code with the config flow. You should be able to finish nonetheless, but without proper validation.

Reply:
Yes, what I can't succeed at the moment is assigning the default values for the Options.
Plus, config flow does not work: it creates everything but probably some parameter is missing because pytuya gets a bad response from the device.

Reply:
Did you fix the for-loop in `_convert_entity` that I mentioned? It uses `self.platform` which you don't set anymore so it won't find any platform specific options. That's probably the problem.

Reply:
I have set it again but still doesn't work. Don't really understand where the default values for missing parameters are assigned, it's really bonkers to me. 😢

Reply:
I will have a look at it.

Shit kinda hit the fan though... In order to implement the schema, I need to pull configuration fields (`flow_schema`) from each platform in `__init__.py`, but since the platform imports stuff from said file we have a circular dependency. We can break things out in a separate file to fix this though, it's not a true circular dependency. I would prefer to finish the CI stuff (fix pydocstyle, flake8 and all that) first as that will help with refactoring. It gives faster feedback since it can tell me if something is misspelled, not imported, etc. So I'm gonna focus on that, so I can do the refactoring and finally we can finish this. Man, this was harder than I thought.

Reply:
OK, I have seen that flow_schema is created correctly and tried to pass it to schema_defaults but for some reason it crashes on 
`    copy = schema.extend({})`
My brain hurts A LOT...